File: 25 programas erlang/programa13.erl
%% --------------------------------------------------------------
%% Módulo: programa13
%% Autor: Van Pratt González Ricardo Adolfo
%% Description: Este programa accede a la llave k de un valor v en
%% un mapa y los imprime
%% Corresponde al ejercicio 13
%%
%% Fecha: 05 de septiembre de 2024
%% --------------------------------------------------------------

-module(programa13).
-export([iterar_mapas/0]).

iterar_mapas() ->
    % Creación del mapa
    MyMap = #{one => 1, "two" => 2.0, <<"three">> => [i, i, i]},
    % Imprime el mapa parte por parte
    maps:fold(
        fun(K, V, ok) ->
            io:format("~p: ~p~n", [K, V])
        end,
        ok,
        MyMap
    ).

File: 50 problemas Prolog/programa13.pl
%% --------------------------------------------------------------
%% Nombre: programa13
%% Autor: Van Pratt González Ricardo Adolfo
%% Description: Este programa crea una lista usando RLE de manera
%% directa.
%%
%% Corresponde al ejercicio 13
%%
%% Fecha: 17 de octubre de 2024
%% --------------------------------------------------------------

% -------- Código en Python (comentado) --------
% def encode_direct(lst):
%     if not lst:
%         return []
%     first = lst[0]
%     n, rest = count(first, lst)
%     return [[n, first]] + encode_direct(rest)

% def count(x, lst):
%     if not lst or lst[0] != x:
%         return 0, lst
%     n, rest = count(x, lst[1:])
%     return n + 1, rest

% if __name__ == "__main__":
%     encoded = encode_direct(['a', 'a', 'b', 'c', 'c', 'c'])
%     print(encoded)
% ----------------------------------------------

% -------- Código en Prolog --------------------
% Codifica directamente una lista usando codificación Run-Length.
% Cuenta cuántos elementos consecutivos son iguales.
encode_direct([], []).
encode_direct([X|Xs], [[N,X]|R]) :- count(X, [X|Xs], N, Rest), encode_direct(Rest, R).

% Cuenta cuántos elementos consecutivos son iguales.
count(_, [], 0, []).
count(X, [X|Xs], N, Rest) :- count(X, Xs, N1, Rest), N is N1 + 1.
count(X, [Y|Ys], 0, [Y|Ys]) :- X \= Y.
% ----------------------------------------------
